### PR TITLE
TSV Downloads of Table Contents

### PR DESCRIPTION
**Pull request by acoffman**
This introduces a new API endpoint and corresponding React component that allows for a user to download a data TSV based on the current parameters used in a query.

I wanted to get feedback on the approach before hooking it up to every table. It also will require some visual polish, I'm not sure how to constrain the size of the MUI loading spinner or error banner, so the button increases in size when either is present. (And the `LoadingButton` appears to be in beta/not shipped with the default library)

The `<TableDownloader/>` component takes two props: `tableName` which is the name of the table you want to download. This is registered in the  `table_resolvers` dictionary in the table downloads controller server side.

This determines which `Resolver` will get used to generate the TSV. (Basically which top level query is this table using).

Then, the resolver in question will need to implement two methods: `table_headers` and `to_row(object:)`. This instructs the resolver how to build the individual TSV rows. For `genes` here I've implemented the same columns we show on the frontend, but this gives us the flexibility to do other things too. We could include columns like ids, or urls to the individual records that we may not want to show visually in the table.

The second prop on the `<TableDownloader>` is `vars` This is simply the same data structure you pass to the `gql` query used to generate the table in the first place. This ensures the TSV is scoped in the same way as the table as the same params will be passed into the resolver. In this examples, it's the gene names used in the initial search query.

Finally, because the TSV can be arbitrarily large, the entire thing is streamed to the client by the server rather than being built up in memory and sent as one large chunk.

Its not a huge change in terms of lines of code, but its pretty dense, so I'm happy to walk through it start to finish on a call!

### CODE REVIEW

**Review by mcannon068nw — APPROVED** [review]
from our call today, implementation seems very sound and amenable to any future updates we would need to do to resolvers, tsv mappings. Going to go ahead and approve pending removal of console log statements and anything else that @katiestahl or @jsstevenson recommend.

For future, we will need to consider how to best style the button still (alignment, etc.) as well as if we want to put a cap in the stream_table method

**Comment by mcannon068nw** [comment]
per meeting today, @jsstevenson or @katiestahl would you be able to look through this PR as well just to make sure everything checks out?

File: server/app/controllers/table_download_controller.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  private
+  def stream_table(resolver: , mapper:)
+    variables = prepare_variables(params[:variables])
+    variables.delete('sortBy')
+    variables.transform_keys! { |k| GraphQL::Schema::Member::BuildType.underscore(k) }
+
+    headers.delete("Content-Length")
+    headers["Cache-Control"] = "no-cache"
+    headers["Content-Type"] = "text/csv"
```
Comment (jsstevenson) [inline server/app/controllers/table_download_controller.rb:28]: we might be able to make this `text/tab-separated-values` but idk if being more specific has any benefits

File: client/src/components/Gene/Categories/CategoryResults.tsx [hunk 2]
```diff
@@ -27,6 +28,7 @@ export const CategoryResults: React.FC = () => {
     <div className="category-results-container">
       <Tabs defaultActiveKey="1" onChange={onChange}>
         <TabPane tab="Unique Matches" key="1">
+        <TableDownloader tableName='gene_category_results' vars={{names: state.searchTerms}}/>
```
Comment (jsstevenson) [inline client/src/components/Gene/Categories/CategoryResults.tsx:31]: There's an upstream issue we need to resolve to enable searching on more fields than just name (and to convert search terms to something unique, like concept ID). Once that's done, we'll need to remember to update this as well.

**Review by jsstevenson — APPROVED** [review]
looks great 👍

**Pull request merged**